### DISCUSSION TITLE
Proposal: Improved Type Checking

### DISCUSSION THREAD

**Discussion by corwinjoy** [cornellius-gp/gpytorch#2180] [Ideas]
@dannyfriar and I have been discussing ideas on how to improve the library. One good suggestion that he made is that stronger type annotations could be a big help. 
In fact, I've noticed that a number of recent issues have involved type difficulties, which stronger annotations might be able to catch. For example:

[Avoid evaluating kernel when adding jitter cornellius-gp/gpytorch#2140](https://github.com/cornellius-gp/gpytorch/pull/2140)
- This turned out to be a broadcasting bug when indexing linear operators.

[The "predict" method in the Deep GP tutorial does not work correctly. cornellius-gp/gpytorch#1892](https://github.com/cornellius-gp/gpytorch/issues/1892)
- The fix was that "The latent variances of multitask DeepGP models are stored in non-interleaved covariance matrices. Previously, the MultitaskMultivariateNormal.marginal method implicitly assumed that the function covariance matrices were interleaved."

Broadly speaking, both of these are kinds of type errors. One with broadcasting and the other with the matrix storage type. 
I think that better checking of [dimensions, element type, matrix storage type] might make the code clearer and help catch some of these bugs.

I am happy to add these annotations, but I am not sure what the best approach would be.
Some options include:
1. [Using TorchTyping](https://github.com/patrick-kidger/torchtyping).
2. [Using PEP 646 / Variadic Generics in Python 3.11](https://peps.python.org/pep-0646).
3. Adding runtime precondition/postcondition asserts.
4. [Other libraries?](https://github.com/patrick-kidger/torchtyping/blob/master/FURTHER-DOCUMENTATION.md#other-libraries-and-resources)
5. Something else?

[Ideas for array shape typing in Python](https://docs.google.com/document/d/1vpMse4c6DrWH5rq2tQSx3qwP_m_0lyn-Ij4WHqQqRHY/edit#heading=h.rkj7d39awayl) gives an overview of how shape types can be helpful.

My plan would be to start with the linear_operator library and see if I can make the functions more explicit there. One simple idea would be to do a minor extension of TorchTyping like:
```
class LinearOperatorType(LinearOperator, TensorTypeMixin):
    base_cls = LinearOperator
```

This would give checking of dimensions + data type but would need to be extended to capture the storage type (dense, interleaved, etc.).

Some of the dynamic functions like slicing and permuting would not be helped much by this approach since it is hard to say too much statically.

Anyway, I wanted to see if there was interest in this idea and/or suggestions on how best to proceed.

Thanks!
Corwin

**Comment by gpleiss** [comment]
Hi Corwin! This is very much something we'd be interested in. I think that TorchTyping would probably the best option, unless there will be some "__future__" mode that gets PEP 646 into earlier versions of Python 3. In general, I think we want to avoid runtime checks because they do add lots of overhead.

I realize our typing is a bit of a mess right now, and so it'd probably also be good to get some linting infrastructure in place that requires that all function arguments have types.

@Balandat thoughts?

**Reply by Balandat** [reply]
I agree that this would be a super useful addition / improvement. `LinearOperator` would definitely be the right place to start.

I think it would probably be good to try and talk to someone who has battle-tested TorchTyping and could speak to how well it works in a complex library before taking the plunge. 

PEP 646 would be great, but we should not require py3.11 for a while. It looks like the design has language syntax backward compatibility in mind (https://peps.python.org/pep-0646/#unpack-for-backwards-compatibility), but I'm not sure to what extent there are good backports. There is https://github.com/python/typing_extensions in the main python org though, so this seems quite promising. Definitely more future-proof than TorchTyping, I would think.

> it'd probably also be good to get some linting infrastructure in place that requires that all function arguments have types.

Yup.

Btw, if you guys are interested in dynamic shapes and a potential alternative for the `NamedTensor`s in pytorch, have you taken a look at https://github.com/facebookresearch/torchdim? This has by now been upstreamed into functorch.

**Reply by jacobrgardner** [reply]
Just to add on, this could be a great opportunity to set the stage for linear_operator's eventual expansion beyond (mostly) positive definite linear operators by baking knowledge of PD linear operators into the type system names already, and annotating ops we know preserve positive definiteness as such by their return types.

**Comment by corwinjoy** [comment]
As an update, I've been exploring this concept using jax typing: [https://github.com/google/jaxtyping](https://github.com/google/jaxtyping)
Why jax typing?
1. The dimension syntax is a bit more expressive than TorchTyping in that it allows dimension arithmetic.
2. The library is more flexible in what types it supports.
3. The author of TorchTyping plans to replace it with the code from jaxtyping.

Please note that:
1. jax typing still needs some customization to remove jax dependencies and add back layout information but this is very doable.
2. In terms of using these signatures, there is no run-time penalty.
3. But, we can *optionally* turn on runtime checking via typeguard or beartype (see jax typing docs).
The main usage would be in the unit tests where we just activate a hook and then we can perform run-time checking of these
signatures to verify correctness and further beef up our tests.

I think these run-time tests are also a helpful debugging tool to track down where dimensions (or storage types such as: dense, sparse, diagnonal, ...) are changing in an unexpected way.

To illustrate some of the challenges, and what this would look like with linear_operator, here are a few key annotations for the base LinearOperator class. I think typing is helpful here but would welcome feedback:

```
# A simple example
def _approx_diagonal(self: Float[LinearOperator, "*batch N N"]) -> Float[torch.Tensor, "*batch N"]:
```

```
# Harder, because the function may take a matrix or 1-D tensor
def _matmul(
    self: Float[LinearOperator, "*batch M N"],
    rhs: Union[Float[torch.Tensor, "*batch N C"], Float[torch.Tensor, "*batch N"]],
) -> Union[Float[torch.Tensor, "*batch M C"], Float[torch.Tensor, "*batch M"]]:
```

```
# Example of a signature where we want to perform math on the dimensions
def cat_rows(
    self: Float[LinearOperator, "*batch M N"],
    cross_mat: Float[torch.Tensor, "*batch O N"],
    new_mat: Float[torch.Tensor, "*batch O N"],
    generate_roots: bool = True,
    generate_inv_roots: bool = True,
    **root_decomp_kwargs,
) -> Float[LinearOperator, "*batch M+O N+O"]:
    """
    Concatenates new rows and columns to the matrix that this LinearOperator represents, e.g.

    .. math::
        \mathbf C = \begin{bmatrix}
            \mathbf A & \mathbf B^\top \\
            \mathbf B & \mathbf D
        \end{bmatrix}

    where :math:`\mathbf A` is the existing LinearOperator, and
    :math:`\mathbf B` (cross_mat) and :math:`\mathbf D` (new_mat)
    are new components. This is most commonly used when fantasizing with
    kernel matrices.
    """
```

Here is a simple example where I think a layout field could help:
```
# current signature for Cholesky
def _cholesky(self, upper: bool = False) -> "TriangularLinearOperator":
```

I think we can improve this by adding a field to jax layout to specify the internal storage layout (like what is done in TorchTyping).
Assuming we use the third field for this, we could define code like:

```
# Linear operator storage layout types
class LinopLayout:
    pass

linop_sparse: LinopLayout
linop_dense: LinopLayout
linop_diagonal: LinopLayout

def _cholesky(
    self: Float[LinearOperator, "*batch N N"], upper: bool = False
) -> Float[LinearOperator, "*batch N N", linop_diagonal]:
```

A big potential application of this could be to track when a linear operator changes from "sparse" to "dense".
Another application might be what @jacobrgardner suggests where we add a field to mark PD status.

Finally, there are some operations that can dynamically add/remove dimensions. I'm not sure how to define a better signature for these:
```
def _expand_batch(self, batch_shape: torch.Size) -> LinearOperator:
"""
Expands along batch dimensions. Return size will be *batch_shape x *matrix_shape.
"""

def _sum_batch(self, dim: int) -> LinearOperator:
"""
Sum the LinearOperator across a batch dimension (supplied as a positive number).
"""
```

Work-in-progress where I am exploring these ideas can be found at https://github.com/corwinjoy/linear_operator/tree/jaxtyping.

**Reply by corwinjoy** [reply]
I'm still working on this. I got run-time type checking turned on for the unit tests which is great in terms of being able to properly test these signatures. The downside is that around 400 of the existing signatures were imprecise/inaccurate so I have to work through a number of changes to get accurate starting signatures. So, this will be a bit of a bigger PR than I would like.  In the meantime, I was at SC22 last week and came across a very cool project by Alexandros Ziogas. He has created a cool framework called DaCe. It can take improved type information and use this to generate better parallel implementations for numerical algorithms. This might help us further accelerate some of our operations once we have these signatures in place. Here is a link - the introductory video is probably the best place to start:
[DaCe Framework](https://github.com/alexnick83/sc22_doctoral_showcase)

**Comment by dannyfriar** [comment]
Awesome, that all sounds great. I'm happy to take a look at the PR when it's ready. DaCe sounds very interesting, I'll check out the video! On 23 Nov 2022 03:04, Corwin Joy ***@***.***> wrote:
I'm still working on this. I got run-time type checking turned on for the unit tests which is great in terms of being able to properly test these signatures. The downside is that around 400 of the existing signatures were imprecise/inaccurate so I have to work through a number of changes to get accurate starting signatures. So, this will be a bit of a bigger PR than I would like.  In the meantime, I was at SC22 last week and came across a very cool project by Alexandros Ziogas. He has created a cool framework called DaCe. It can take improved type information and use this to generate better parallel implementations for numerical algorithms. This might help us further accelerate some of our operations once we have these signatures in place. Here is a link - the introductory video is probably the best place to start:
DaCe Framework

**Comment by corwinjoy** [comment]
@dannyfriar @gpleiss  I have gone ahead and done an initial PR with these improved signatures. I would say that there is more to do along these lines but I believe this is a step in the right direction.  Anyway, I hope it can create a productive discussion around these ideas.

Thanks!
